src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,273 @@
+            if (is_object($value)) {
+                $this->set[$key] = clone $value;
+            }
+        }
+    }
+
+    /**
+     * Constructs a lazy BSON array.
+     *
+     * @param PackedArray|array|null $input An input for a lazy array.
+     *        When given a BSON array, this is treated as input. For lists
+     *        this constructs a new BSON array using fromPHP.
+     */
+    public function __construct($input = null, ?CodecLibrary $library = null)
+    {
+        if ($input === null) {
+            $this->bson = PackedArray::fromPHP([]);
+        } elseif ($input instanceof PackedArray) {
+            $this->bson = $input;
+        } elseif (is_array($input)) {
+            $this->bson = PackedArray::fromPHP(array_values($input));
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+        }
+
+        $this->library = $library;
+    }
+
+    public function bsonSerialize(): array
+    {
+        // Always use LazyBSONArrayCodec for BSON serialisation
+        $codec = new LazyBSONArrayCodec();
+        $codec->attachLibrary($this->getLibrary());
+
+        // @psalm-suppress InvalidReturnStatement
+        return $codec->encode($this)->toPHP();
+    }
+
+    /** @return Iterator<int, mixed> */
+    public function getIterator(): Iterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON array
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        $seen = [];
+
+        // Use AsArrayIterator to ensure we're indexing from 0 without gaps
+        return new AsListIterator(
+            new CallbackIterator(
+                new CallbackFilterIterator(
+                    $itemIterator,
+                    /** @param mixed $value */
+                    function ($value, int $offset) use (&$seen): bool {
+                        // Skip keys that were unset or handled in a previous iterator
+                        return ! isset($this->unset[$offset]) && ! isset($seen[$offset]);
+                    }
+                ),
+                /**
+                 * @param mixed $value
+                 * @return mixed
+                 */
+                function ($value, int $offset) use (&$seen) {
+                    // Mark key as seen, skipping any future occurrences
+                    $seen[$offset] = true;
+
+                    // Return actual value (potentially overridden by offsetSet)
+                    return $this->offsetGet($offset);
+                }
+            )
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this);
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        $offset = (int) $offset;
```

Comment:
This is one of those moments where I hate PHP arrays. If the key is numeric, it will be converted to an int, see https://3v4l.org/PuaTC. It seems to me that this should return if `! is_numeric($offset)` and cast to `int` otherwise.

Reply:
If offset is a string and not exactly an integer, it remains a string. Consider https://3v4l.org/C4O78#v8.2.5:

```php
<?php

function test(array|ArrayAccess $a) {
    $tests = [
        true,
        false,
        (int) 2.5,
        '3.5',
        '4.0',
        '5.',
        '6',
        7.0,
        8.5,
    ];

    foreach ($tests as $key) {
        $a[$key] = $key;
    }
    
    var_dump($a);
}

test([]);
test(new ArrayObject);
```

The native array and ArrayObject behave the same, so I'll spare you the full output:

```
Deprecated: Implicit conversion from float 8.5 to int loses precision in /in/C4O78 on line 17
array(9) {
  [1]=>
  bool(true)
  [0]=>
  bool(false)
  [2]=>
  int(2)
  ["3.5"]=>
  string(3) "3.5"
  ["4.0"]=>
  string(3) "4.0"
  ["5."]=>
  string(2) "5."
  [6]=>
  string(1) "6"
  [7]=>
  float(7)
  [8]=>
  float(8.5)
}
```

The explicit cast in `offsetExists()` circumvents the deprecation notice for loss of precision.

It also unconditionally converts strings to an integer, which is inconsistent with some of the test cases above.